### ISSUE TITLE
yate segfaults

### ISSUE THREAD

**Issue by clehner** [openwrt/telephony#99]
Original issue at https://github.com/openwrt/packages/issues/2323

yate (5.5.0-1-1)
OpenWrt Designated Driver r48135
TP-Link TL-WDR3500

When I start yate or attempt to run it from the command line, I get a segfault immediately. I have tried with yate modules installed, and with no modules at all installed. Even when passing --help, it crashes before it gets to process command-line arguements.

Kernel log output:

```
do_page_fault(): sending SIGSEGV to yate for invalid read access from 00000000
epc = 77416ce0 in libc.so[773a0000+92000]
ra  = 772c919b in libyate.so.5.5.0[77286000+e9000]
```

gdb shows that the crash is in musl's ~~tss_set~~ tss_get function, but can't get the rest of the call stack:

```
$ gdb -q yate
Reading symbols from yate...(no debugging symbols found)...done.
(gdb) r
Starting program: /usr/bin/yate

Program received signal SIGSEGV, Segmentation fault.
0x77fc8ce0 in tss_get () from /lib/ld-musl-mips-sf.so.1
(gdb) bt
#0  0x77fc8ce0 in tss_get () from /lib/ld-musl-mips-sf.so.1
warning: GDB can't find the start of the function at 0x77e7b19a.

    GDB is unable to find the start of the function at 0x77e7b19a
and thus can't determine the size of that function's stack frame.
This means that GDB may be unable to access that stack frame, or
the frames below it.
    This problem is most likely caused by an invalid program counter or
stack pointer.
    However, if you think GDB should simply search farther back
from 0x77e7b19a for code which looks like the beginning of a
function, you can increase the range of the search using the `set
heuristic-fence-post' command.
#1  0x77e7b19b in ?? ()
(gdb) set heuristic-fence-post 100000
(gdb) bt
#0  0x77fc8ce0 in tss_get () from /lib/ld-musl-mips-sf.so.1
warning: GDB can't find the start of the function at 0x77e7b19a.
#1  0x77e7b19b in ?? ()
```

~~tss_set in musl:
http://git.musl-libc.org/cgit/musl/tree/src/thread/tss_set.c~~

~~I couldn't find any code calling tss_set in musl, yate, or uclibcxx (a dependency of yate).~~ I am trying to compile a debug build of yate now. Is anyone able to reproduce this issue or have an idea of why it might be happening?

**Comment by jow-** [comment]
Maybe something is doing pthread_* calls before having pthreading initialized? There was a similar issue in motion, see https://github.com/openwrt/packages/pull/1847

Its just a guess though.

**Comment by jow-** [comment]
Additional note: `tss_get()` might get called through `pthread_getspecific()`

**Comment by clehner** [comment]
@jow- That sounds like it could be it. When my toolchain finishes compiling I will try to find out where yate is calling one of those during startup.

**Comment by jow-** [comment]
The code in `thread.cpp` looks suspcious, it implements a `ThreadPrivateKeyAlloc keyAllocator` which wraps `pthread_key_create()` but it does not appear to be called anywhere. I'm totally clueless about C++ so I'm not sure if that statement alone already calls the constructor of `ThreadPrivateKeyAlloc`.

Even if it does, gcc is maybe optimizing the statement away...

**Comment by jow-** [comment]
Maybe a case of this: http://stackoverflow.com/questions/1229430/how-do-i-prevent-my-unused-global-variables-being-compiled-out

**Comment by jow-** [comment]
I wonder if this change might be enough already:

```diff
--- a/engine/Thread.cpp
+++ b/engine/Thread.cpp
@@ -120,7 +120,7 @@ public:
     }
 };
 
-static ThreadPrivateKeyAlloc keyAllocator;
+ThreadPrivateKeyAlloc keyAllocator;
 #endif /* _WINDOWS */
 
 static TokenDict s_prio[] = {
```

**Comment by clehner** [comment]
Could be. I will try it once my build environment is ready

**Comment by clehner** [comment]
Ok, I got it to compile but the change making keyAllocator non-static doesn't seem to help.
I tried recompiling with `CONFIG_DEBUG=y` and `CONFIG_TARGET_OPTIMIZATION="-O0"` to see if that would reveal a backtrace in gdb, but it didn't.

I was however able to isolate the crash (using printf statements) to the call to pthread_getspecific (as you mentioned) in Thread.cpp. And ThreadPrivateKeyAlloc is not getting called before that. Also, main (in main.cpp) hasn't been called yet.

Changing the keyAllocator line to
`ThreadPrivateKeyAlloc *keyAllocator = new ThreadPrivateKeyAlloc();` did not help either.
